### PR TITLE
bpo-35022: Add `__fspath__` support to `unittest.mock.MagicMock`

### PR DESCRIPTION
**Pull request by maxbelanger**
This is a naive implementation of what has been working for us internally. 

https://bugs.python.org/issue35022

### CODE REVIEW

File: Doc/library/unittest.mock.rst
Comment (vstinner) [inline Doc/library/unittest.mock.rst:1702]: You should document the change using ".. versionchanged:: 3.8 (describe your change on the next line)" (search in the doc for other examples using verisonchanged).

File: Lib/unittest/test/testmock/testmagicmethods.py
Comment (vstinner) [inline Lib/unittest/test/testmock/testmagicmethods.py]: Please add: ``self.assertTrue(m.__fspath__.assert_called_once())``

**Review by vstinner — APPROVED** [review]
LGTM. But I would prefer to have a second review, @mariocj89?

**Review by mariocj89 — APPROVED** [review]
LGTM! Feels a little "weird" but I guess there is no better way, I've added some minor suggestions in case you find them useful @vstinner

File: Lib/unittest/mock.py
Comment (mariocj89) [inline Lib/unittest/mock.py]: This will elide all naming information. See:
```
>>> os.fspath(m.absolute)
'MagicMock/140083495922656'
>>> str(m.absolute)
"<MagicMock name='mock.absolute' id='140083495922656'>"
```

Or when a mock is given a name. The id is unique per attribute and probably quite unuseful for the user when debugging issues.


Have you considered:
`lambda self: f"{self._extract_mock_name()}/{id(self)}"`

It would look nicer in some scenarios like:

```
>>> os.fspath(mock.MagicMock(**{"__fspath__.return_value": "abc"}).absolute)
'mock.absolute/140476946856592'
```

But maybe weird for some others like:
```
>>> os.fspath(mock.MagicMock() / 'other' / 'path')
'mock.__truediv__().__truediv__()/140476946879824'
```

Comment (mariocj89) [inline Lib/unittest/mock.py]: missing trailing comma.

Add a comma at the end of this line to ease reviews that add items in the future.


`'__fspath__': lambda self: "%s/%s" % (type(self).__name__, id(self))` -> `'__fspath__': lambda self: "%s/%s" % (type(self).__name__, id(self)),`

Comment (maxbelanger) [reply Lib/unittest/mock.py]: Agree that naming information should be preserved, what would you think of:
```
lambda self: f"{type(self).__name__}/{self._extract_mock_name()}/{id(self)}"
```

The `__truediv__` case is a bit strange but it is correct, though.

**Comment by vstinner** [comment]
@mariocj89: does it look better to you?

**Comment by mariocj89** [comment]
Sure! I think it will be easier to debug. :)

**Comment by vstinner** [comment]
> Sure! I think it will be easier to debug. :)

I mean: can you please review the updated PR? Like approving it if you consider that it now looks good to you?

**Review by mariocj89 — APPROVED** [review]
Much better indeed! :)

**Pull request merged**
